File: .indio-backups/server-ts-20260507-002053/src/core/context.ts
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import type { AppConfig } from "../config.js";
import type { CalendarAdapter } from "../adapters/calendar.js";
import type { WeatherAdapter } from "../adapters/weather.js";
import type { ContextBundle, TriggerSource, UserProfile } from "@indio/contracts";
import { StateStore } from "./state.js";

export class ContextService {
  constructor(
    private readonly config: AppConfig,
    private readonly state: StateStore,
    private readonly weather: WeatherAdapter,
    private readonly calendar: CalendarAdapter
  ) {}

  async loadProfile(): Promise<UserProfile> {
    const [taste, routines, moodRules, playlistsJson] = await Promise.all([
      readFile(resolve(this.config.userDir, "taste.md"), "utf8"),
      readFile(resolve(this.config.userDir, "routines.md"), "utf8"),
      readFile(resolve(this.config.userDir, "mood-rules.md"), "utf8"),
      readFile(resolve(this.config.userDir, "playlists.json"), "utf8")
    ]);

    return {
      taste,
      routines,
      moodRules,
      playlists: JSON.parse(playlistsJson) as UserProfile["playlists"]
    };
  }

  async build(params: { source: TriggerSource; userInput?: string }): Promise<ContextBundle> {
    const [systemPrompt, profile, weather, calendar] = await Promise.all([
      readFile(this.config.promptPath, "utf8"),
      this.loadProfile(),
      this.weather.getSnapshot(),
      this.calendar.getEventsForToday()
    ]);

    return {
      systemPrompt,
      profile,
      weather,
      calendar,
      recentMessages: this.state.listRecentMessages(6),
      recentPlays: this.state.listRecentPlays(4),
      currentTime: new Date().toISOString(),
      source: params.source,
      userInput: params.userInput
    };
  }

  async getTasteSummary(): Promise<{
    tasteHighlights: string[];
    routineHighlights: string[];
    playlists: UserProfile["playlists"];
  }> {
    const profile = await this.loadProfile();

    const toHighlights = (value: string) =>
      value
        .split("\n")
        .map((line) => line.trim())
        .filter((line) => line.startsWith("-"))
        .map((line) => line.replace(/^-+\s*/, ""))
        .slice(0, 4);

    return {
      tasteHighlights: toHighlights(profile.taste),
      routineHighlights: toHighlights(profile.routines),
      playlists: profile.playlists
    };
  }
}

File: server/app/core/context.py
from __future__ import annotations

import json

from app.adapters.calendar import CalendarAdapter
from app.adapters.weather import WeatherAdapter
from app.config import AppConfig
from app.core.state import StateStore
from app.models import ContextBundle, TriggerSource, UserProfile, utc_now_iso


class ContextService:
    def __init__(
        self,
        config: AppConfig,
        state: StateStore,
        weather: WeatherAdapter,
        calendar: CalendarAdapter,
    ) -> None:
        self.config = config
        self.state = state
        self.weather = weather
        self.calendar = calendar

    async def load_profile(self) -> UserProfile:
        taste = (self.config.user_dir / "taste.md").read_text(encoding="utf-8")
        routines = (self.config.user_dir / "routines.md").read_text(encoding="utf-8")
        mood_rules = (self.config.user_dir / "mood-rules.md").read_text(encoding="utf-8")
        playlists = json.loads((self.config.user_dir / "playlists.json").read_text(encoding="utf-8"))
        return UserProfile(taste=taste, routines=routines, mood_rules=mood_rules, playlists=playlists)

    async def build(self, *, source: TriggerSource, user_input: str | None = None) -> ContextBundle:
        profile = await self.load_profile()
        return ContextBundle(
            system_prompt=self.config.prompt_path.read_text(encoding="utf-8"),
            profile=profile,
            weather=await self.weather.get_snapshot(),
            calendar=await self.calendar.get_events_for_today(),
            recent_messages=self.state.list_recent_messages(6),
            recent_plays=self.state.list_recent_plays(4),
            current_time=utc_now_iso(),
            source=source,
            user_input=user_input,
        )

    async def get_taste_summary(self) -> dict[str, object]:
        profile = await self.load_profile()

        def highlights(value: str) -> list[str]:
            return [
                line.removeprefix("-").strip()
                for line in value.splitlines()
                if line.strip().startswith("-")
            ][:4]

        return {
            "tasteHighlights": highlights(profile.taste),
            "routineHighlights": highlights(profile.routines),
            "playlists": [playlist.model_dump(mode="json", by_alias=True) for playlist in profile.playlists],
        }
